staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1166,6 +1166,137 @@ func TestCacherSendBookmarkEvents(t *testing.T) {
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            fmt.Sprintf("pod-%d", index),
+				Namespace:       "ns",
+				ResourceVersion: fmt.Sprintf("%v", 100+index),
+			},
+		}
+	}
+
+	var currentResourceVersion uint64 = 3
+	createdPods := []*example.Pod{
+		makePod(1),
+		makePod(2),
+		makePod(currentResourceVersion),
+	}
+	for i := range createdPods {
+		if err := cacher.watchCache.Add(createdPods[i]); err != nil {
+			t.Fatalf("failed to add a pod: %v", err)
+		}
+	}
+
+	expectedWatchEventsFunc := func(createdPods []*example.Pod) []watch.Event {
+		var expectedWatchEvents []watch.Event
+		for _, pod := range createdPods {
+			expectedWatchEvents = append(expectedWatchEvents, watch.Event{Type: watch.Added, Object: pod})
+		}
+		return expectedWatchEvents
+	}
+
+	trueVal, falseVal := true, false
+
+	scenarios := []struct {
+		name                string
+		allowWatchBookmarks bool
+		sendInitialEvents   *bool
+		expectedWatchEvents []watch.Event
+	}{
+		{
+			name:                "allowWatchBookmarks=false, sendInitialEvents=false",
+			allowWatchBookmarks: false,
+			sendInitialEvents:   &falseVal,
+			expectedWatchEvents: expectedWatchEventsFunc(createdPods),
+		},
+		{
+			name:                "allowWatchBookmarks=false, sendInitialEvents=true",
+			allowWatchBookmarks: false,
+			sendInitialEvents:   &trueVal,
+			expectedWatchEvents: expectedWatchEventsFunc(createdPods),
+		},
+		{
+			name:                "allowWatchBookmarks=true, sendInitialEvents=true",
+			allowWatchBookmarks: true,
+			sendInitialEvents:   &trueVal,
+			expectedWatchEvents: expectedWatchEventsFunc(createdPods),
+		},
+		{
+			name:                "allowWatchBookmarks=true, sendInitialEvents=false",
+			allowWatchBookmarks: true,
+			sendInitialEvents:   &falseVal,
+			expectedWatchEvents: expectedWatchEventsFunc(createdPods),
+		},
+		{
+			name:                "allowWatchBookmarks=false, sendInitialEvents=nil",
+			allowWatchBookmarks: true,
+			expectedWatchEvents: expectedWatchEventsFunc(createdPods),
+		},
+	}
+	for i := range scenarios {
+		if scenarios[i].allowWatchBookmarks && scenarios[i].sendInitialEvents != nil && *scenarios[i].sendInitialEvents {
+			scenarios[i].expectedWatchEvents = append(scenarios[i].expectedWatchEvents, watch.Event{
+				Type: watch.Bookmark,
+				Object: &example.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						ResourceVersion: strconv.FormatUint(100+currentResourceVersion, 10),
+						Annotations:     map[string]string{metav1.InitialEventsAnnotationKey: "true"},
+					},
+				},
+			})
+		}
+
+		pred := storage.Everything
+		pred.AllowWatchBookmarks = scenarios[i].allowWatchBookmarks
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+		w, err := cacher.Watch(ctx, "pods/ns", storage.ListOptions{ResourceVersion: "100", SendInitialEvents: scenarios[i].sendInitialEvents, Predicate: pred})
+		if err != nil {
+			t.Fatalf("Failed to create watch: %v", err)
+		}
+
+		timeout := time.NewTicker(10 * time.Second)
```

Comment:
verifyEvents(t, w, expectedEvents, true)
testCheckNoMoreResults(t, w)

Reply:
yes, it looks like tidy. In order to reuse `testCheckNoMoreResults` method, I make it to `TestCheckNoMoreResults`